libvoikko/typescript/libvoikko.d.ts

```diff
@@ -0,0 +1,198 @@
+export = Libvoikko;
+
+declare class Libvoikko {
+  constructor();
+  init(language: string): Voikko;
+}
+
+declare interface GrammarError {
+  errorCode: number;
+  startPos: number;
+  errorLen: number;
+  suggestions: Array<string>;
+  shortDescription: string;
+}
+
+declare interface Analysis {
+  BASEFORM: string;
+  CLASS: string;
+  FSTOUTPUT: string;
+  NUMBER: string;
+  SIJAMUOTO: string;
+  STRUCTURE: string;
+  WORDBASES: string;
+}
+
+declare interface Token {
+  type: TokenType;
+  text: string;
+}
+
+type TokenType = 'WORD' | 'WHITESPACE' | 'PUNCTUATION';
+
+declare interface Sentence {
+  text: string;
+  nextStartType: NextStartType;
+}
+
+type NextStartType = 'PROBABLE' | 'NONE';
+
+type SuggestionStrategy = 'OCR' | 'TYPO';
+
+declare interface Voikko {
+  /**
+   * Releases the resources allocated by libvoikko for this instance.
+   */
+  terminate: () => void;
+
+  /**
+   * Check the spelling of given word. Return true if the word is correct,
+   * false if it is incorrect.
+   */
+  spell: (word: string) => boolean;
+
+  /**
+   * a list of suggested spellings for given (misspelled) word. If the given
+   * word is correct, the list contains only the word itself
+   */
+  suggest: (word: string) => Array<string>;
+
+  /**
+   * Check the given text for grammar errors and return a list of
+   * GrammarError objects representing the errors that were found.
+   * Unlike the C based API this method accepts multiple paragraphs
+   * separated by newline characters.
+   */
+  grammarErrors: (text: string, language?: string) => Array<GrammarError>;
+
+  /**
+   * Analyze the morphology of given word and return the list of analysis results.
+   */
+  analyze: (word: string) => Array<Analysis>;
+
+  /**
+   * Split the given natural language text into a list of Token objects.
+   */
+  tokens: (text: string) => Array<Token>;
+
+  /**
+   * Split the given natural language text into a list of Sentence objects.
+   */
+  sentences: (text: string) => Array<Sentence>;
+
+  /**
+   * Return a character pattern that describes the hyphenation of given word.
+   *
+   * ' ' = no hyphenation at this character,
+   * '-' = hyphenation point (character at this position is preserved in the hyphenated form),
+   * '=' = hyphentation point (character at this position is replaced by the hyphen.)
+   */
+  getHyphenationPattern: (word: string) => string;
+
+  /**
+   * Returns the given word in fully hyphenated form.
+   */
+  hyphenate: (word: string) => string;
```

Comment:
Missing optional parameters:
```suggestion
  hyphenate: (word: string, separator?: string, allowContextChanges?: boolean) => string;
```